### PR TITLE
Support for cupyx.scipy.linalg

### PR DESCRIPTION
**Pull request by quasiben**
- [x] Closes #https://github.com/dask/dask/issues/7537
- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask` / `isort dask`

### CODE REVIEW

File: dask/array/linalg.py [hunk 1]
```diff
@@ -1188,14 +1184,18 @@ def _key(i, j):
...
                 )
 
     graph = HighLevelGraph.from_collections(name, dsk, dependencies=[a, b])
+
+    a_meta = meta_from_array(a)
+    b_meta = meta_from_array(b)
     res = _solve_triangular_lower(
-        np.array([[1, 0], [1, 2]], dtype=a.dtype), np.array([0, 1], dtype=b.dtype)
+        np.array([[1, 0], [1, 2]], dtype=a.dtype, like=a_meta),
+        np.array([0, 1], dtype=b.dtype, like=b_meta),
```
Comment (jrbourbeau) [inline dask/array/linalg.py:1198]: Perhaps we should use `array_safe` here for smoothly support the `like=` keyword?

https://github.com/dask/dask/blob/865515ca94b0da2872413d5d38b8ecf1e6b9fe1c/dask/array/utils.py#L467

cc @pentschev

Comment (quasiben) [reply dask/array/linalg.py:1198]: Yes!  I had meant to come back to that. Thanks for the reminder @jrbourbeau

File: dask/array/utils.py [hunk 2]
```diff
@@ -565,6 +565,22 @@ def svd_flip(u, v, u_based_decision=False):
     return u, v
 
 
+################
+# scipy.linalg #
+################
```
Comment (pentschev) [inline dask/array/utils.py:570]: No other functions in this file have this comment, it makes me think `_is_nep18_active` is part of `scipy.linalg`. I would remove this.

File: dask/array/utils.py [hunk 3]
```diff
@@ -565,6 +565,22 @@ def svd_flip(u, v, u_based_decision=False):
...
+
+def solve_triangular_safe(a, b, lower=False):
+    import scipy.linalg
+
+    if is_cupy_type(a):
+        import cupyx.scipy.linalg
+
+        return cupyx.scipy.linalg.solve_triangular(a, b, lower=lower)
+
+    return scipy.linalg.solve_triangular(a, b, lower=lower)
```
Comment (pentschev) [inline dask/array/utils.py:581]: There are other `scipy.linalg` functions that could use the same idea. Could we generalize this?

I haven't checked them, but if the first argument is always an ndarray, we could do something like the following:

```python
def scipy_linalg_safe(func_name, *args, **kwargs):
    if is_cupy_type(a):
        import cupyx.scipy.linalg
        func = getattr(cupyx.scipy.linalg, func_name)
    else:
        import scipy.linalg
        func = getattr(scipy.linalg, func_name)

    return func(*args, **kwargs)
```

And the caller would then obviously call `scipy_linalg_safe("solve_triangular", a, b, lower=lower)`.

Comment (quasiben) [reply dask/array/utils.py:615]: I think this is a good idea generally but at the moment dask, I think, only uses three functions from scipy.linalg: `solve_triangluar`, `lu`, `cholesky`.  `solve_traingluar` is currently supported in `cupyx.scipy.linalg` and `cholesky` is from cupy.linalg.  In my last commit I added support for `cholesky`

File: dask/array/utils.py [hunk 4]
```diff
@@ -427,12 +427,36 @@ def zeros_like_safe(a, shape, **kwargs):
...
+    """
+    if like is None:
+        return np.zeros(*args, **kwargs)
+    else:
+        try:
+            return np.zeros(*args, like=meta_from_array(like), **kwargs)
+        except TypeError:
+            return np.zeros(*args, **kwargs)
+
+
```
Comment (pentschev) [inline dask/array/utils.py:459]: Why is this necessary? We have `zeros_like_safe` that handles the exact same case. IOW, these two are equivalent:

```
np.zeros_like(a, shape=(b, c))
np.zeros((b, c), like=a)
```

File: dask/array/linalg.py [hunk 5]
```diff
@@ -1321,7 +1324,7 @@ def _cholesky(a):
     for i in range(vdim):
         for j in range(hdim):
             if i < j:
-                dsk[name, i, j] = (np.zeros, (a.chunks[0][i], a.chunks[1][j]))
+                dsk[name, i, j] = (_zeros_safe, (a.chunks[0][i], a.chunks[1][j]), a)
```
Comment (pentschev) [inline dask/array/linalg.py:1327]:
```suggestion
                dsk[name, i, j] = (zeros_like_safe, a, shape=(a.chunks[0][i], a.chunks[1][j]))
```

Comment (pentschev) [reply dask/array/linalg.py:1327]: This should work the same way without the need for `_zeros_safe`.

File: dask/array/linalg.py [hunk 6]
```diff
@@ -1351,7 +1354,9 @@ def _cholesky(a):
 
     graph_upper = HighLevelGraph.from_collections(name_upper, dsk, dependencies=[a])
     graph_lower = HighLevelGraph.from_collections(name, dsk, dependencies=[a])
-    cho = scipy.linalg.cholesky(np.array([[1, 2], [2, 5]], dtype=a.dtype))
+    a_meta = meta_from_array(a)
+    cho = np.linalg.cholesky(array_safe([[1, 2], [2, 5]], dtype=a.dtype, like=a_meta))
+    # cho = scipy.linalg.cholesky(np.array([[1, 2], [2, 5]], dtype=a.dtype))
```
Comment (pentschev) [inline dask/array/linalg.py:1359]:
```suggestion
```

File: dask/array/utils.py [hunk 7]
```diff
@@ -427,12 +427,36 @@ def zeros_like_safe(a, shape, **kwargs):
     using the old behavior, returning ``np.zeros(shape, **kwargs)``.
     """
     try:
+        foo = np.zeros_like(a, shape=shape, **kwargs)
+        print("foo: ", type(foo))
```
Comment (pentschev) [inline dask/array/utils.py:431]:
```suggestion
```

File: dask/array/utils.py (hunk 3 above)
Comment (pentschev) [reply dask/array/utils.py:615]: Might still be worth doing this for `lu` later, as both are part of `cupyx.scipy.linalg`.

File: dask/array/utils.py [hunk 8]
```diff
@@ -565,6 +589,32 @@ def svd_flip(u, v, u_based_decision=False):
...
+    if is_cupy_type(a):
+        import cupy
+
+        xp = cupy
+
+    chunk = xp.matmul(a, b)
+    # Since we have performed the contraction via matmul
+    # but blockwise expects all dimensions back, we need
+    # to add one dummy dimension back
+    return chunk[..., xp.newaxis]
```
Comment (pentschev) [inline dask/array/utils.py:604]: I think this would be better left at its original place in `routines._matmul`, as this seems like a specialization from that function specifically. IOW, we only need `return xp.matmul(a, b)` here after NumPy/CuPy has been correctly chosen.

**Review by pentschev — CHANGES_REQUESTED** [review]

File: dask/array/utils.py (hunk 4 above)
Comment (quasiben) [reply dask/array/utils.py:459]: You're right.  However, during graph construction I think we can only provide args (as opposed to kwargs). What do you think about keeping `_zeros_safe` and pointing it at `zeros_like_safe`.  This makes it possible to do the following:

> (_zeros_safe, a, (a.chunks[0][i], a.chunks[1][j]), meta_from_array(a))

File: dask/array/utils.py (hunk 3 above)
Comment (quasiben) [reply dask/array/utils.py:615]: Ah, I was looking at CuPy 8.  `lu` is in CuPy 9 so I'll add this in as you suggested

File: dask/array/utils.py (hunk 4 above)
Comment (pentschev) [reply dask/array/utils.py:459]: With `zeros_like_safe` the `shape` argument is the second positional argument, making it perfectly possible to just do:

```python
(zeros_like_safe, a, (a.chunks[0][i], a.chunks[1][j]))
```

or

```python
(zeros_like_safe, meta_from_array(a), (a.chunks[0][i], a.chunks[1][j]))
```

Not sure which of the above you actually need here.

**Review by jrbourbeau — COMMENTED** [review]
Just wanted to check in here. Is this something we'd like to get into the release tomorrow, or are we okay waiting until the one after that?

**Comment by quasiben** [comment]
No, this can wait. But thanks for checking in

File: dask/array/core.py [hunk 9]
```diff
@@ -286,6 +286,7 @@ def dotmany(A, B, leftfunc=None, rightfunc=None, **kwargs):
         A = map(leftfunc, A)
     if rightfunc:
         B = map(rightfunc, B)
+    breakpoint()
```
Comment (jrbourbeau) [inline dask/array/core.py:289]: I suspect this is left over from developing locally 

```suggestion
```

Comment (quasiben) [reply dask/array/core.py]: d'oh!  Thanks @jrbourbeau !

File: dask/array/utils.py [hunk 10]
```diff
@@ -565,6 +565,41 @@ def svd_flip(u, v, u_based_decision=False):
...
+    # for gpu/cpu checking
+    a = args[0]
+    if is_cupy_type(a):
+        import cupyx.scipy.linalg
+
+        func = getattr(cupyx.scipy.linalg, func_name)
+    else:
+        import scipy.linalg
+
+        func = getattr(scipy.linalg, func_name)
```
Comment (jakirkham) [inline dask/array/utils.py:594]: I wonder if we could use `Dispatch` for this kind of thing

https://github.com/dask/dask/blob/c5633c21e754df6b2c13195f1e2688e347c48963/dask/sizeof.py#L17-L28

We could also then do lazy registration with this

https://github.com/dask/dask/blob/c5633c21e754df6b2c13195f1e2688e347c48963/dask/sizeof.py#L97-L106

Comment (quasiben) [reply dask/array/utils.py:594]: I think that's a good idea if this grows more.  As it stands now, this PR only adds `solve_triangular`.  I tried adding in `lu` but ran into a number of issues to due to lack of `int64` support in the cupy implementation

Comment (jakirkham) [reply dask/array/utils.py:594]: FWIW I had some code here that might be relevant ( https://github.com/jakirkham/dask/commit/81adbc82c484436a33d4617eecf1b7f0eb84bf1b ). Though feel free to skip it if you would rather

File: dask/array/utils.py (hunk 8 above)
Comment (pentschev) [reply dask/array/utils.py:580]: @quasiben could you check my comment above?

Comment (quasiben) [reply dask/array/utils.py]: Sorry, I missed this @pentschev  .  I addressed this in https://github.com/dask/dask/pull/7563/commits/b13833c82eaf5ff470eff066ce1e427b91fd1001 .

File: dask/array/utils.py (hunk 10 above)
Comment (jakirkham) [reply dask/array/utils.py:579]: Also thinking about this in the context of dispatch refactoring work going on around Arrays and DataFrames ( https://github.com/dask/dask/pull/7505 ) ( https://github.com/dask/dask/pull/7503 )

**Review by pentschev — APPROVED** [review]
Overall LGTM, thanks @quasiben .

File: dask/array/utils.py (hunk 10 above)
Comment (jakirkham) [reply dask/array/utils.py:579]: >  I tried adding in `lu` but ran into a number of issues to due to lack of `int64` support in the cupy implementation

Just to clarify is that suppose to be `float64`? While we may be able to provide `int64` arrays, expect these would likely need to be casted to `float64` and then the result would also be in `float64`. Though maybe I'm misunderstanding something here

**Comment by jsignell** [comment]
Seems like this is good to merge right @quasiben?

File: dask/array/utils.py (hunk 10 above)
Comment (quasiben) [reply dask/array/utils.py:579]: No, I think we need int64 support for `lu` functionality.  Or maybe a conversion is what scipy lu does and I'm unaware of it

**Comment by quasiben** [comment]
Yeah, this can be merged now
